### PR TITLE
Support annotation for nullable values in a container

### PR DESCRIPTION
**Pull request by nicklan**
When arrow fixes https://github.com/apache/arrow-rs/issues/6391, our code will start to fail since we don't allow for nulls in partition values, but they can be null.

This adds an annotation that will cause the generated schema to allow null values. We already have the semantics that if value is null on the arrow side, then `materialize` will simply not include it in the final map, so nothing else needs to change.

Tested by:
1. All tests still pass
2. Added a new test for the `add` schema that ensures it sets the `value_contains_null` field to true on the `partitionValues` map.
3. Looking at generated code and noting we get
```rust
HashMap::<
  String,
  String,
>::get_nullable_container_struct_field("partitionValues"),
```

### CODE REVIEW

File: derive-macros/src/lib.rs [hunk 1]
```diff
@@ -74,7 +82,16 @@ fn gen_schema_fields(data: &Data) -> TokenStream {
                         _ => panic!("Can only handle <> type path args"),
                     }
                 });
-                quote_spanned! { field.span() => #(#type_path_quoted),* get_struct_field(stringify!(#name))}
+                if have_schema_null {
+                    if let Some(first_ident) = type_path.path.segments.first().map(|seg| &seg.ident) {
+                        if first_ident != "HashMap" && first_ident != "Vec" {
+                            panic!("Can only use schema_container_values_null on HashMap or Vec fields, not {first_ident:?}");
```
Comment (scovich) [inline derive-macros/src/lib.rs:88]: Why would it be correct to delete null entries from a Vec? It would change the indexes of all entries after it, which could cause a problem if the indexes were somehow meaningful?

(I also wonder what potential use cases that even serves?)

File: kernel/src/actions/mod.rs [hunk 2]
```diff
@@ -141,6 +141,7 @@ pub struct Add {
     pub path: String,
 
     /// A map from partition column to value for this logical file.
+    #[schema_container_values_null]
```
Comment (scovich) [inline kernel/src/actions/mod.rs:144]: We should consider `//` commenting why the attribute is needed? 
(if I understand correctly, partition values can be physically null, and the attribute causes the read to correctly handle that even tho the resulting HashMap will simply not contain them)

However, the above explanation is a bit hard to parse, and doesn't immediately follow from the attribute. That makes me wonder if we need to take a different approach to avoid trouble. 

For example -- should the attribute be called `drop_null_container_values`? Then we could document it as "I know the underlying data for this non-nullable map/vec is nullable -- drop null values instead of erroring out"

File: derive-macros/src/lib.rs [hunk 3]
```diff
@@ -84,8 +89,8 @@ fn gen_schema_fields(data: &Data) -> TokenStream {
                 });
                 if have_schema_null {
                     if let Some(first_ident) = type_path.path.segments.first().map(|seg| &seg.ident) {
-                        if first_ident != "HashMap" && first_ident != "Vec" {
-                            panic!("Can only use schema_container_values_null on HashMap or Vec fields, not {first_ident:?}");
+                        if first_ident != "HashMap" {
+                            panic!("Can only use drop_null_container_values on HashMap fields, not {first_ident:?}");
```
Comment (scovich) [inline derive-macros/src/lib.rs:93]: IIRC `panic` technically works, because it just kills the macro invocation and produces a compiler error... but IIRC there's a better way, emitting an error to the output token stream so the span info can give a more precise error message?

**Review by scovich — APPROVED** [review]

File: derive-macros/src/lib.rs (hunk 3 above)
Comment (nicklan) [reply derive-macros/src/lib.rs:93]: Yeah, we could use [`to_compile_error()`](https://docs.rs/syn/latest/syn/struct.Error.html)

I've created https://github.com/delta-incubator/delta-kernel-rs/issues/348 to convert to this

Comment (zachschuermann) [reply derive-macros/src/lib.rs:93]: TIL!

**Review by zachschuermann — APPROVED** [review]
lgtm!
